File: 백준/Gold/10942. 팰린드롬？/팰린드롬？.js
const fs = require("fs")
const filePath = process.platform === "linux" ? "dev/stdin" : "백준/input.txt"
const input = fs.readFileSync(filePath).toString().split("\n")

const N = Number(input[0])
const nums = input[1].split(" ").map(Number)
const M = Number(input[2])

const dp = Array.from(Array(N), () => new Array(N).fill(0))

for (let i = 0; i < N; i++) {
  dp[i][i] = 1
}

for (let j = 0; j < N; j++) {
  for (let i = 0; i < j; i++) {
    let [S, E] = [i, j]
    if (nums[S] !== nums[E]) {
      continue
    }
    S += 1
    E -= 1
    if (S > E || dp[S][E]) dp[i][j] = 1
  }
}

let answer = []
for (let i = 3; i < M + 3; i++) {
  const [S, E] = input[i].split(" ").map(Number)
  answer.push(dp[S - 1][E - 1])
}
console.log(answer.join("\n"))

File: 백준/Gold/10942. 팰린드롬？/팰린드롬？.py
import sys
input = sys.stdin.readline

N = int(input())
nums = list(map(int, input().split()))
M = int(input())
questions = [list(map(int, input().split())) for _ in range(M)]

dp = [[0] * N for _ in range(N)]
for i in range(N):
    dp[i][i] = 1

for j in range(N):
    for i in range(0, j+1):
        S, E = i, j
        if nums[i] != nums[j]:
            continue
        S += 1
        E -= 1
        if S > E or dp[S][E]:
            dp[i][j] = 1

for S, E in questions:
    print(dp[S-1][E-1])
